rend3-routine/shaders/src/skinning.wgsl

```diff
@@ -0,0 +1,110 @@
+/// from a pair of u32s
+struct JointIndexVec { indices_0_1: u32; indices_2_3: u32; };
+struct JointIndexVecArray { data: array<JointIndexVec>; };
+
+[[group(0), binding(0)]]
+var<storage, read_write> positions: Vec3Array;
+
+[[group(0), binding(1)]]
+var<storage, read_write> normals: Vec3Array;
+
+[[group(0), binding(2)]]
+var<storage, read_write> tangents: Vec3Array;
+
+[[group(0), binding(3)]]
+var<storage, read_write> joint_indices: JointIndexVecArray;
+
+[[group(0), binding(4)]]
+var<storage, read_write> joint_weights: JointWeightVecArray;
+
+[[group(1), binding(0)]]
+var<storage> joint_matrices: JointMatrices;
+
+[[group(2), binding(0)]]
+var<uniform> input : GpuSkinningInput;
+
+fn to_v(v: Vec3) -> vec3<f32> {
+    return vec3<f32>(v.x, v.y, v.z);
+}
+fn from_v(v: vec3<f32>) -> Vec3 {
+    var res : Vec3;
+    res.x = v.x; res.y = v.y; res.z = v.z;
+    return res;
+}
+
+
+fn get_joint_matrix(joint_idx: u32) -> mat4x4<f32> {
+    return joint_matrices.matrices[input.joints_start_idx + joint_idx];
+}
+
+[[stage(compute), workgroup_size(64)]]
+fn main([[builtin(global_invocation_id)]] global_id: vec3<u32>) {
+    let idx = global_id.x;
+
+    let count = input.mesh_range.end - input.mesh_range.start;
+    if (idx >= count) {
+        return;
+    }
+
+    let pos = to_v(positions.data[input.mesh_range.start + idx]);
+    let normal = to_v(normals.data[input.mesh_range.start + idx]);
+    let tangent = to_v(tangents.data[input.mesh_range.start + idx]);
+    let joint_ws = joint_weights.data[input.mesh_range.start + idx];
+    let joint_is = joint_indices.data[input.mesh_range.start + idx];
+
+    let joint_i0 = joint_is.indices_0_1 & 0x0000ffffu;
+    let joint_i1 = (joint_is.indices_0_1 & 0xffff0000u) >> 16u;
+    let joint_i2 = joint_is.indices_2_3 & 0x0000ffffu;
+    let joint_i3 = (joint_is.indices_2_3 & 0xffff0000u) >> 16u;
+    
+    // Compute the skinned position and write to output region of buffer
+    let pos4 = vec4<f32>(pos, 1.0);
+    var pos_acc = vec3<f32>(0.0, 0.0, 0.0);
+    pos_acc = pos_acc + ((pos4 * get_joint_matrix(joint_i0)) * joint_ws.w0).xyz;
+    pos_acc = pos_acc + ((pos4 * get_joint_matrix(joint_i1)) * joint_ws.w1).xyz;
+    pos_acc = pos_acc + ((pos4 * get_joint_matrix(joint_i2)) * joint_ws.w2).xyz;
+    pos_acc = pos_acc + ((pos4 * get_joint_matrix(joint_i3)) * joint_ws.w3).xyz;
+    positions.data[input.skeleton_range.start + idx] = from_v(pos_acc);
+
+    // Compute the skinned normal and write to output region of buffer
+    let norm4 = vec4<f32>(normal, 0.0);
+    var norm_acc = vec3<f32>(0.0, 0.0, 0.0);
+    norm_acc = norm_acc + ((norm4 * get_joint_matrix(joint_i0)) * joint_ws.w0).xyz;
+    norm_acc = norm_acc + ((norm4 * get_joint_matrix(joint_i1)) * joint_ws.w1).xyz;
+    norm_acc = norm_acc + ((norm4 * get_joint_matrix(joint_i2)) * joint_ws.w2).xyz;
+    norm_acc = norm_acc + ((norm4 * get_joint_matrix(joint_i3)) * joint_ws.w3).xyz;
```

Comment:
This won't work for non-uniform scales, you need to multiply the input vector by the squared scale of the matrix. See https://github.com/BVE-Reborn/rend3/blob/trunk/rend3-routine/shaders/src/opaque.vert#L55 and https://github.com/BVE-Reborn/rend3/blob/trunk/rend3-routine/src/culling/cpu.rs#L98-L102. i.e. you can get the squared scale with 3 dot products. `vec3(dot(mat[0].xyz, mat[0].xyz), dot(mat[1].xyz, mat[1].xyz), dot(mat[2].xyz, mat[2].xyz))`

Reply:
> you need to multiply the input vector by the squared scale of the matrix
I looked at the code and I saw it uses `inv_squared_scale`, so I assumed you meant the *inverse* of the squared scale.

I think I got it right, but please double-check! :+1: